feat: Add PostgreSQL advisory lock for migrations

shortstacked opened the pull request:
## Summary

Ensures only one n8n instance runs database migrations and initializes auth roles service in multi-main setups to prevent race conditions.
Implements transaction-level advisory locks for PostgreSQL to avoid orphaned locks and improve reliability.
Retries lock acquisition with timeout and provides detailed error message with instructions for manual intervention.
Disables advisory lock for SQLite databases.
Runs main instances in parallel during testing to improve efficiency.

## Related Linear tickets, Github issues, and Community forum posts

<!--
Include links to **Linear ticket** or Github issue or Community forum post.
Important in order to close *automatically* and provide context to reviewers.
https://linear.app/n8n/issue/
-->
<!-- Use "closes #<issue-number>", "fixes #<issue-number>", or "resolves #<issue-number>" to automatically close issues when the PR is merged. -->


## Review / Merge checklist

- [ ] PR title and summary are descriptive. ([conventions](../blob/master/.github/pull_request_title_conventions.md)) <!--
   **Remember, the title automatically goes into the changelog.
   Use `(no-changelog)` otherwise.**
-->
- [ ] [Docs updated](https://github.com/n8n-io/n8n-docs) or follow-up ticket created.
- [ ] Tests included. <!--
   A bug is not considered fixed, unless a test is added to prevent it from happening again.
   A feature is not complete without tests.
-->
- [ ] PR Labeled with `release/backport` (if the PR is an urgent fix that needs to be backported)

shortstacked committed:
feat: Add PostgreSQL advisory lock for migrations

Ensures only one n8n instance runs database migrations and
initializes auth roles service in multi-main setups to prevent
race conditions.

Implements transaction-level advisory locks for PostgreSQL to
avoid orphaned locks and improve reliability.

Retries lock acquisition with timeout and provides detailed
error message with instructions for manual intervention.

Disables advisory lock for SQLite databases.

Runs main instances in parallel during testing to improve efficiency.

shortstacked committed:
Merge branch 'master' into pg-advisory-lock

tomi reviewed (commented):
This has still one issue: It does not work when pool size is configured to be one:

```
➜ DB_TYPE=postgresdb DB_POSTGRESDB_PASSWORD=postgres DB_POSTGRESDB_POOL_SIZE=1 DB_LOGGING_ENABLED=true DB_LOGGING_OPTIONS=all node bin/n8n
Initializing n8n process
query: SELECT * FROM current_schema()
query: CREATE EXTENSION IF NOT EXISTS "uuid-ossp"
query: SELECT version();
n8n ready on ::, port 5678
query: START TRANSACTION
query: SELECT pg_try_advisory_xact_lock(1850)
Database connection timed out
Database connection timed out
503 Database is not ready!
query: ROLLBACK
There was an error running database migrations
timeout exceeded when trying to connect
query: SELECT 1
query: SELECT 1
query: SELECT 1
Database connection recovered
```

This is because the transaction for the advisory lock reserves one DB connection, and there's no connection left in the pool for the transactions. So we would need to move this to the library side if we wanted to use a trx level lock.

This is also why the Postgres tests timed out. We should not allow merging of the PR in case a test times out. I created [CAT-1985](https://linear.app/n8n/issue/CAT-1985/pr-should-not-be-mergeable-if-tests-timeout) for this.

So I would suggest that we revert back to the session level lock with the timeout. In the long run we should move this locking logic into our typeorm fork.

shortstacked committed:
Switches to session-level advisory lock

Uses session-level advisory locks instead of transaction-level locks for PostgreSQL.

This change avoids the need to maintain a transaction during the lock's lifetime, addressing potential issues with single connection pool setups. It also simplifies lock management, enabling explicit release and improving reliability.
